open-craft/learner-profile-dashboard
# Collect and and store quantitative data for adaptive engine

cf. [OC-4386](https://tasks.opencraft.com/browse/OC-4386)

This PR adds logic to create or update `Score` records for answers to individual options of quantitative questions. Scores are only stored if:

* A given answer option `influences_recommendations`.
* The answer option is linked to a knowledge component.

**Test instructions**

1. Run migrations.

1. Assuming that the DB of your LPD installation has the test data from #3, load additional test data by `%cpaste`ing commands from `commands-quantitative.txt` (attached to the ticket).

1. Access the LPD in the LMS and provide answers to all quantitative questions, then click "Submit your answers".

1. Run the Django shell and make sure that `Score` records were created for all answer options belonging to MCQ, MRQ, and Ranking questions:

    ```
    from lpd.models import *

    Score.objects.count()
    # Should output 12
    ```

1. Print some details about individual `Score` records to verify that the data matches the answers that you provided earlier. Note that the `value` of a given `Score` does not need to be identical to the raw answer that you provided; see `MultipleChoiceQuestion._get_score` and `RankingQuestion._get_score` for details on the transformations that are performed on answer values to produce scores:

    ```
    for score in Score.objects.all(): print(score.learner, score.knowledge_component, score.knowledge_component.answer_option, score.value)
    # Should output something like this:
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 8: mr_kc4, MR KC 4>, <AnswerOption: AnswerOption 8: Something else:>, 0.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 7: mr_kc3, MR KC 3>, <AnswerOption: AnswerOption 7: high>, 0.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 6: mr_kc2, MR KC 2>, <AnswerOption: AnswerOption 6: medium>, 1.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 5: mr_kc1, MR KC 1>, <AnswerOption: AnswerOption 5: low>, 1.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 10: rank_kc2, Ranking KC 2>, <AnswerOption: AnswerOption 10: I love salad>, 0.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 11: rank_kc3, Ranking KC 3>, <AnswerOption: AnswerOption 11: I love meat>, 0.3333333333333333)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 9: rank_kc1, Ranking KC 1>, <AnswerOption: AnswerOption 9: I love candy>, 1.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 12: rank_kc4, Ranking KC 4>, <AnswerOption: AnswerOption 12: I love:>, 0.6666666666666666)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 2: mc_kc2, MC KC 2>, <AnswerOption: AnswerOption 2: Blue>, 1.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 3: mc_kc3, MC KC 3>, <AnswerOption: AnswerOption 3: Green>, 0.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 1: mc_kc1, MC KC 1>, <AnswerOption: AnswerOption 1: Red>, 1.0)
    # (<User: 7wJN637PYRIpN4kkaW5CEg++>, <KnowledgeComponent: KnowledgeComponent 4: mc_kc4, MC KC 4>, <AnswerOption: AnswerOption 4: Your favourite color:>, 1.0)
    ```

1. Change some of your answers and click "Submit your answers" again, then repeat the previous two steps. The number of `Score` records should be unchanged, but the `value` of each `Score` belonging to an answer you changed should reflect the changes that you made.

**Notes**

- Test coverage increased to 98% overall :)

**Reviewers**
